tools/autograd/derivatives.yaml

```diff
@@ -772,6 +785,12 @@
 - name: mul.Tensor(Tensor self, Tensor other) -> Tensor
   self: mul_tensor_backward(grad, other, self.scalar_type())
   other: mul_tensor_backward(grad, self, other.scalar_type())
+  result: other_t * self_p.conj() + self_t * other_p.conj()
+
+- name: mul_.Tensor(Tensor(a!) self, Tensor other) -> Tensor(a!)
+  self: mul_tensor_backward(grad, other, self.scalar_type())
+  other: mul_tensor_backward(grad, self, other.scalar_type())
+  result: self_t.mul_(other_p.conj()).add_(other_t * (self_p / other_p).conj())
```

Comment:
This is a lot better than what we had before, but I'm still not looking forward to having to duplicate formulas for *every* in place definition, just so that the inplace forward AD can be faster. For one, the backwards formulas are always going to be the same, so at the very least you should make it so that you don't have to copy paste them. But more generally, it seems like it should be possible to derive these automatically (although we don't have the machinery to do so right now).

Reply:
I'm a little confused by the `other_t * (self_p / other_p).conj()` term here, how come `self_p` has to be divided by `other_p` here?

Reply:
The problem is that because this is inplace, `self` in the formulas (forward and backward) actually contain the result.
I can change the name here to improve readability if the codegen handles it properly.

Reply:
> The problem is that because this is inplace, self in the formulas (forward and backward) actually contain the result.

I mean, I acknowledge that, computationally, inplace can get you in trouble (as you describe). But I still think of derivatives.yaml as being a *mathematical* specification of what derivatives are supposed to be, and the compilers job is to figure out how to translate it into executable code.